### PR TITLE
feat: New `Upload` scalar type and  local file upload support for `createMediaItem`

### LINKED ISSUE #311: createMediaItem doesn't allow you to upload from a local file
We need to debug why the createMediaItem mutation isn't allowing for uploads from a local file.

### PR DESCRIPTION
**Pull request by renatonascalves**
What does this implement/fix? Explain your changes.
---------------------------------------------------
- Introduces a new `Upload` scalar type
- The `createMediaItem` mutation is updated to use this new scalar type

Does this close any currently open issues?
------------------------------------------
fixes #311 

Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------

Here is a snippet which can be used for confirmation (created using Paw):
```
## Local File
curl -X "POST" "https://test.site/graphql" \
     -H 'Content-Type: multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__' \
     -F "0=" \
     -F "operations={\"query\":\"mutation($file: Upload) {\\n  createMediaItem(\\n    input: {\\n      file: $file\\n      title: \\\"Awesome Image\\\"\\n      clientMutationId: \\\"CreateThatImage\\\"\\n    }\\n  ) {\\n    clientMutationId\\n    mediaItem {\\n      id\\n      title\\n      sourceUrl\\n    }\\n  }\\n}\\n\",\"variables\":{\"file\":null},\"operationName\":null}" \
     -F "map={ \"0\": [\"variables.file\"] }"
```

Screen Shot 2022-08-21 at 21 05 53

Any other comments?
-------------------
- This pull request was created from the #2482, since the PHPCS changes should be merged first, as soon as that happens, a lot of the commits will disappear.
- After this pull request is code reviewed and approved, documentation will need to be updated to reflect this new scalar type
- Also, more unit test support should be added to account for edge cases

Where has this been tested?
---------------------------
**Operating System:** MacOS Monterey 12.4
**WordPress Version:** 6.0.1
**PHP Version:** 7.4

### CODE REVIEW

**Comment by renatonascalves** [comment]
Creating pull request as a draft while #2482 is not merged to avoid confusion. :)

**Comment by justlevine** [comment]
It's hard to tell with #2482 included - is this a copy and/or compatible with [WPGraphQL Upload](https://github.com/dre1080/wp-graphql-upload)?

If the former, we should probably attribute it.

If the latter, we probably want to auto-deactivate that plugin and provide a notice that it's now in core and the plugin can be deleted.

If neither, we either need to handle type collisions, or save this PR for a breaking release.

**Comment by jasonbahl** [comment]
@renatonascalves this is exciting! 

I'll try and review this week

**Comment by renatonascalves** [comment]
@justlevine 

> is this a copy and/or compatible with [WPGraphQL Upload](https://github.com/dre1080/wp-graphql-upload)?

Yes! With a few changes but it was highly inspired by that plugin.

> f neither, we either need to handle type collisions nor save this PR for a breaking release (edit: and coordinate with @dre1080 ).

I'd prefer to leave this for a breaking release (allowing time for plugin authors to update their plugins) instead of trying to account for that external plugin (those external plugins).

> we should probably link/attribute it in the file headers.

About attribution, I wonder if the Shout outs section is not a better place for that: https://github.com/wp-graphql/wp-graphql#shout-outs

**Comment by justlevine** [comment]
@renatonascalves 

> I'd prefer to leave this for a breaking release (allowing time for plugin authors to update their plugins) instead of trying to account for that external plugin (those external plugins) within WP GraphQL.

If that's indeed the approach you want to take, then realistically this probably won't be merged for a while**. 

There's a lot of high-priority breaking changes on the docket, and tons of code-debt that can/should be addressed before them. While breaking changes dont _need_ to be scary or saved for a single release, when the GraphQL schema is involved, they can stop the entire app from building, and we dont have a semver aware mechanism to prevent autoupdate from potentially wrecking who knows how many sites.

**speaking for myself, who has no actual decision making power ;-) 

Now that #2482 is merged, I'll do a code audit shortly 🤞

File: src/Server/WPHelper.php [hunk 1]
```diff
@@ -46,22 +50,79 @@ public function parseRequestParams( $method, array $bodyParams, array $queryPara
 		 * persisted queries (and ends up being a bit more flexible than
 		 * graphql-php's built-in persistentQueryLoader).
 		 *
-		 * @param array $data An array containing the pieces of the data of the GraphQL request
-		 * @param array $request_context An array containing the both body and query params
+		 * @param array  $parsed_body_params An array containing the pieces of the data of the GraphQL request.
+		 * @param array  $request_context    An array containing the both body and query params.
+		 * @param string $method             The method of the request (GET, POST, etc).
 		 */
+		$parsed_body_params = apply_filters( 'graphql_request_data', $parsed_body_params, $request_context, $method );
```
Comment (justlevine) [inline src/Server/WPHelper.php:57]: Not sure if this is a flaw with the _old_ code (using the filter to modify `$parsed_body_params` and `$parsed_query_params`), but this change means `$parsed_query_params` doesnt get a filter at all.

File: src/Server/WPHelper.php [hunk 2]
```diff
@@ -46,22 +50,79 @@ public function parseRequestParams( $method, array $bodyParams, array $queryPara
...
+	 * @param mixed $json Json.
+	 * @return mixed
+	 */
+	private function decode_json( $json ) {
+
+		if ( ! is_string( $json ) ) {
+			return $json;
+		}
+
+		return json_decode( stripslashes( $json ), true );
```
Comment (justlevine) [inline src/Server/WPHelper.php:123]: We should probably check `json_last_error()`, instead of just assuming we're going to get an array.

File: src/Type/Scalar/Upload.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	 * Register the scalar Upload type.
+	 *
+	 * @return void
+	 */
+	public static function register_type() {
+		register_graphql_scalar(
+			'Upload',
+			[
+				'description'  => __( 'The `Upload` special type represents a file to be uploaded in the same HTTP request as specified by [graphql-multipart-request-spec](https: //github.com/jaydenseric/graphql-multipart-request-spec).', 'wp-graphql' ),
+				'serialize'    => fn() => static::serialize(),
```
Comment (justlevine) [inline src/Type/Scalar/Upload.php:34]: arrow functions are only supported in PHP 7.4 or later.  These should be changed to long-form

File: src/Registry/TypeRegistry.php [hunk 4]
```diff
@@ -337,6 +338,11 @@ public function init_type_registry( TypeRegistry $type_registry ) {
 		PostObjectUnion::register_type( $this );
 		TermObjectUnion::register_type( $this );
 
+		/**
+		 * Register scalar.
+		 */
+		Upload::register_type();
```
Comment (justlevine) [inline src/Registry/TypeRegistry.php:344]: We should probably register scalars at the top of this method, to allow any of the types preceding it.

Comment (justlevine) [reply src/Registry/TypeRegistry.php:344]: I'm assuming stanning is failing because Composer's autoloader hasnt been updated. `composer dump-autoload --no-dev --optimize` should fix that.

File: tests/wpunit/MediaItemMutationsTest.php [hunk 5]
```diff
@@ -756,6 +756,83 @@ public function testCreateMediaItemMutation() {
...
+		// Set our local file.
+		$this->create_variables['input']['file']     = $this->file;
+		$this->create_variables['input']['slug']     = 'test-medium';
+		$this->create_variables['input']['filePath'] = null;
+
+		/**
+		 * Create the createMediaItem
+		 */
+		$actual = $this->createMediaItemMutation();
+
```
Comment (justlevine) [inline tests/wpunit/MediaItemMutationsTest.php:777]: https://github.com/wp-graphql/wp-graphql/actions/runs/3033573100/jobs/4930710142#step:7:64742

once the bug itself if fixed, we should `$this->assertArrayNotHasKey( 'errors', $actual )`

**Review by justlevine — CHANGES_REQUESTED** [review]
Tests are failing. I left comments where appropriate.

**Comment by renatonascalves** [comment]
@justlevine  I had a change of heart and will update this pull request to account for the code review suggestions and a type collisions. :)

File: src/Server/WPHelper.php (hunk 1 above)
Comment (renatonascalves) [reply src/Server/WPHelper.php:57]: The `$request_context` holds the data from both body and query parameters. I remember not adding a different hook only because this data is already available here.

Comment (renatonascalves) [reply src/Server/WPHelper.php:57]: Is it still worth adding it in a separate hook? Reading the code, I didn't see how useful this would be introducing a new hook before this action is run.

File: src/Server/WPHelper.php (hunk 2 above)
Comment (renatonascalves) [reply src/Server/WPHelper.php:123]: Good idea!

File: src/Type/Scalar/Upload.php (hunk 3 above)
Comment (renatonascalves) [reply src/Type/Scalar/Upload.php:34]: Good catch!

File: tests/wpunit/MediaItemMutationsTest.php [hunk 6]
```diff
@@ -756,6 +764,117 @@ public function testCreateMediaItemMutation() {
...
+								'sourceUrl' => wp_get_attachment_image_src( $attachment_id, 'medium_large' )[0],
+							],
+						],
+					],
+				],
+			],
+		];
+
+		$this->assertEquals( $expected, $actual['data'] );
+	}
```
Comment (kidunot89) [inline tests/wpunit/MediaItemMutationsTest.php:876]: @renatonascalves Maybe use the `assertQuerySuccessful()` assertion, and the `expectField()`, `expectedObject()`, `expectedNode()`, and `expectedEdge()` to strictly test the parts of the query you need to test.

**Review by kidunot89 — CHANGES_REQUESTED** [review]
@renatonascalves Looks great bro, I don't have too much to nitpick. You even met the WPCS compliances. See my comments, especially the ones about the unit test 👌🏿. Great work, this has been a long time coming 🥂.

File: src/Server/WPHelper.php [hunk 7]
```diff
@@ -46,26 +49,122 @@ public function parseRequestParams( $method, array $bodyParams, array $queryPara
 		 * persisted queries (and ends up being a bit more flexible than
 		 * graphql-php's built-in persistentQueryLoader).
 		 *
-		 * @param array $data An array containing the pieces of the data of the GraphQL request
-		 * @param array $request_context An array containing the both body and query params
+		 * @param array  $request_data    An array containing the pieces of the data of the GraphQL request.
+		 * @param array  $request_context An array containing the both body and query params.
+		 * @param string $method          The method of the request (GET, POST, etc).
 		 */
+		$filtered_resquest_data = apply_filters( 'graphql_request_data', $request_data, $request_context, $method );
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply src/Server/WPHelper.php:56]: Good catch!

File: src/Server/WPHelper.php [hunk 8]
```diff
@@ -46,26 +49,122 @@ public function parseRequestParams( $method, array $bodyParams, array $queryPara
...
+		}
+
+		$map    = $this->decode_json( $parsed_body_params['map'] );
+		$result = $this->decode_json( $parsed_body_params['operations'] );
+
+		foreach ( $map as $fileKey => $locations ) {
+			$items = &$result;
+
+			foreach ( $locations as $location ) {
+				foreach ( explode( '.', $location ) as $key ) {
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply src/Server/WPHelper.php:111]: I think so. Let me review this again.

File: src/Type/Scalar/Upload.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	 * @return void
+	 */
+	public static function register_type( TypeRegistry $type_registry ) {
+		$type_registry->register_scalar(
+			'Upload',
+			[
+				'description'  => __( 'The `Upload` special type represents a file to be uploaded in the same HTTP request as specified by [graphql-multipart-request-spec](https: //github.com/jaydenseric/graphql-multipart-request-spec).', 'wp-graphql' ),
+				'serialize'    => function () {
+					static::serialize();
+				},
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply src/Type/Scalar/Upload.php:43]: Interesting! Let me look at that.

File: tests/wpunit/MediaItemMutationsTest.php (hunk 6 above)
Comment (renatonascalves) [reply tests/wpunit/MediaItemMutationsTest.php:876]: Good suggestion! I was trying to follow the same pattern already used in the test file, though.

**Comment by renatonascalves** [comment]
Closing this based on this [feedback](https://wp-graphql.slack.com/archives/CDE8Y3MN3/p1688671804766179?thread_ts=1688669186.063079&cid=CDE8Y3MN3).

**Comment by justlevine** [comment]
@renatonascalves in the interim WPGraphQL Upload 0.1.5 is out with some sanitization fixes.
